src/goto-programs/goto_contractor.cpp

```diff
@@ -228,39 +603,44 @@ ibex::Ctc *goto_contractort::create_contractor_from_expr2t(const expr2tc &expr)
   return contractor;
 }
 
-bool goto_contractort::is_constraint_operator(const expr2tc &expr)
+bool expr_to_ibex_parser::is_constraint_operator(const expr2tc &expr)
 {
   expr2tc e = get_base_object(expr);
   return (is_comp_expr(e) && !is_notequal2t(e)) || is_arith_expr(e);
 }
 
+bool expr_to_ibex_parser::is_constraint_operator_not(const expr2tc &expr)
+{
+  expr2tc e = get_base_object(expr);
+  return (is_comp_expr(e) && !is_equality2t(e)) || is_arith_expr(e);
+}
+
 ibex::NumConstraint *
-goto_contractort::create_constraint_from_expr2t(const expr2tc &expr)
+expr_to_ibex_parser::create_constraint_from_expr2t(const expr2tc &expr)
 {
-  ibex::NumConstraint *c;
+  ibex::NumConstraint *c = nullptr;
+  auto base_object = get_base_object(expr);
 
   if(is_unsupported_operator_in_constraint(expr))
   {
     parse_error(expr);
     return nullptr;
   }
+
   if(is_not2t(expr))
   {
-    //TODO: implement "not" flag  and process.
-    parse_error(expr);
-    return nullptr;
+    c = create_constraint_from_expr2t_not(to_not2t(base_object).value);
+    return c;
   }
 
-  const relation_data &rel =
-    dynamic_cast<const relation_data &>(*get_base_object(expr));
-
+  auto rel = &dynamic_cast<const relation_data &>(*get_base_object(expr));
   ibex::Function *f, *g;
-  f = create_function_from_expr2t(rel.side_1);
-  g = create_function_from_expr2t(rel.side_2);
+  f = create_function_from_expr2t(rel->side_1);
+  g = create_function_from_expr2t(rel->side_2);
   if(f == nullptr || g == nullptr)
     return nullptr;
-  //TODO: implement "not" flag  and process.
-  switch(get_base_object(expr)->expr_id)
+
+  switch(base_object->expr_id)
   {
   case expr2t::expr_ids::greaterthanequal_id:
     c = new ibex::NumConstraint(*vars, (*f)(*vars) >= (*g)(*vars));
```

Comment:
My understanding was that ibex used smart pointers. This doesn't seem to be the case. I will fix this one.

Reply:
fixed. I created lists that keep track of each created object and delete it in the deconstructor.

Reply:
Sorry, but this memory "management" code has to be fixed properly:
- You have already been dynamically allocating (with `new`) all these ibex objects, which may or may not be smart pointers; please check! If they indeed are, do not dynamically allocate them.
- Now you're remembering which ones to delete in a **linked list**. This is just not great at all, because each element of a linked list is dynamically allocated **again**. So now you're performing 2 dynamic allocations where I believe none are necessary.

Why do I believe that none are necessary? Because their API, as used by the contractors code, takes ibex (sub-)objects by reference, not by (raw) pointer. That's why you're dereferencing the dynamically allocated ibex objects all the time you're using their API. Please correct me if I'm wrong, but this doesn't seem to be right usage.

I realize that it's been there as it is, and this PR is already big enough, so we can postpone this to another PR, but it should be the next step for the contractors code.

Reply:
I've had a quick look at the ibex implementation. Basically all of its types assume pointer stability for contained objects. That's certainly not an idiomatic choice, but it explains why allocations of the form the contractors code is doing are indeed necessary.

They even have their own type for tracking allocations similar to your change: https://github.com/ibex-team/ibex-lib/blob/master/src/tools/ibex_Memory.h It also uses linked list, and in addition even an dynamically allocated sub-object to store in there. Which is worse. But it has a useful comment in `~Memory()`:
> deleted in reverse order (in case some objects depend on previous objects)

I presume the only way to use ibex' strange API efficiently in a dynamic setting is to write an own arena-style memory allocator ontop of a [std::deque](https://en.cppreference.com/w/cpp/container/deque) or similar. @ all: Am I missing something here?

Anyway, so I agree we need to track the allocated objects. My suggestion as long as we don't have this arena-style memory allocator would be to append to a std::vector instead of prepending to std::list to track the pointers. For deletion the std::vector needs to be traversed in reverse insertion order.

Reply:
Thank you, Franz. Your comments gave me great insight. I have changed the lists to vectors and traversed them in reverse. May I ask you why vectors instead of lists? is it to avoid more dynamic allocations?

Reply:
> May I ask you why vectors instead of lists? is it to avoid more dynamic allocations?

Happy to help. Yes, it is to avoid dynamic allocations. A recent profiling of ESBMC already shows that we spend a lot of time in malloc and grouping allocations together is a good way to keep the pressure down. Especially when we can't exactly know how complex the expressions are going to be. A std::vector only ever keeps one array, regardless how many elements it holds. The array is reallocated roughly for every log(n) added element.

> I have changed the lists to vectors and traversed them in reverse.

The `~Contractor` destructor still seems to go through in forward direction, though. I wonder why both destructors delete the allocated ibex elements...

Reply:
> The ~Contractor destructor still seems to go through in forward direction, though. I wonder why both destructors delete the allocated ibex elements...

Fixed. Sorry, I missed that one.